File: Problems/No.1338_Reduce_Array_Size_to_The_Half/reduce_array_to_the_half.cpp
class Solution {
public:
    int minSetSize(vector<int>& arr) {
        
        // parameters
        int removed_set_num = 0;
        std::unordered_map<int, int> IntSize;
        std::priority_queue<std::pair<int, int>> IntSetSize;

        // start computing
        // calculate the size of each integer & order them
        for (auto &n: arr) IntSize[n]++;
        for (auto &[Int, Size]: IntSize) IntSetSize.push({Size, Int});

        // removing them iteratively
        int total_size = arr.size();
        int removed_size = 0;
        while (removed_size < total_size/2) {

            removed_set_num++;
            removed_size += IntSetSize.top().first;
            IntSetSize.pop();

        }

        return removed_set_num;

    }
};

File: Problems/No.1338_Reduce_Array_Size_to_The_Half/reduce_array_to_the_half.py
class Solution(object):
    def minSetSize(self, arr):
        """
        :type arr: List[int]
        :rtype: int
        """

        # parameters
        removed_set_num = 0
        IntSize = {}
        IntSetSize = []

        # start computing
        # calculate the size of each integer & order them
        for n in arr:
            IntSize[n] = IntSize.get(n, 0) + 1
        for [Int, Size] in IntSize.items():
            heapq.heappush(IntSetSize, [(-1)*Size, Int])

        # removing them iteratively
        removed_size = 0
        while ((-1)*removed_size < len(arr)//2):
            removed_set_num += 1
            removed_size += heapq.heappop(IntSetSize)[0]

        return removed_set_num
